File: Olympiad/COCI/official/2009/contest4_solutions/periodni.cpp
#include <cstdio>

#define MAXN 500
#define MAXK 500
#define MOD 1000000007

typedef long long llint;

int gcd( int a, int b ) { return b == 0 ? a : gcd( b, a%b ); }

llint komb[MAXN+1][MAXK+1];
llint povrh[MAXK+1];

struct node { llint dp[MAXK+1]; } nodes[3*MAXN], *allocator = nodes;

int k, h[MAXN];

void calc_povrh( int n ) {
   static int gore[MAXK], i;
   povrh[0] = 1;
   for( i = 1; i <= k && n >= i; ++i ) {
      gore[i-1] = n - i+1;
      
      int dole = i;
      for( int j = 0; j < i; ++j ) {
         int g = gcd( gore[j], dole );
         dole /= g;
         gore[j] /= g;
      }
      povrh[i] = 1;
      for( int j = 0; j < i; ++j ) povrh[i] = povrh[i] * gore[j] % MOD;
   }
   while( i <= k ) povrh[i++] = 0;
}

node *rec( int a, int b, int h0 ) {
   node *curr = allocator++;

   int leaf = 1, minh = 1000000000, minh_index = -1;
   for( int i = a; i < b; ++i ) {
      if( h[i] < minh ) { minh = h[i]; minh_index = i; }
      if( h[i] != h[a] ) leaf = 0;
   }

   if( leaf ) {
      calc_povrh( minh-h0 );  
      
      for( int i = 0; i <= k; ++i )
         curr->dp[i] = komb[b-a][i] * povrh[i]%MOD;
      
   } else if( minh > h0 ) {
      node *up = rec( a, b, minh );

      calc_povrh( minh-h0 );

      for( int i = 0; i <= k; ++i ) {
         curr->dp[i] = 0;
         for( int j = 0; j <= i && b-a-j >= 0; ++j )
            curr->dp[i] += up->dp[j] * povrh[i-j]%MOD * komb[b-a-j][i-j]%MOD;
         curr->dp[i] %= MOD;
      }
   } else {
      node *left = rec( a, minh_index, h0 );
      node *right = rec( minh_index+1, b, h0 );
      
      for( int i = 0; i <= k; ++i ) {
         curr->dp[i] = 0;
         for( int j = 0; j <= i; ++j ) 
            curr->dp[i] = curr->dp[i] + left->dp[j] * right->dp[i-j]%MOD;
         curr->dp[i] %= MOD;
      }
   }
   return curr;
}

int main( void ) {
   int n;
   scanf( "%d%d", &n, &k );
   for( int i = 0; i < n; ++i ) scanf( "%d", &h[i] );

   for( int i = 0; i <= n; ++i ) {
      komb[i][0] = 1;
      for( int j = 1; j <= i; ++j ) 
         komb[i][j] = komb[i][j-1] * (i-j+1) % MOD;
   }

   node * root = rec( 0, n, 0 );
   printf( "%d\n", root->dp[k] );

   return 0;
}

File: Olympiad/COCI/official/2009/contest4_solutions/periodni.pas
program periodni;

const MAXN = 500;
const MAXK = 500;
const MODC = 1000000007;

function gcd(a, b : longint) : longint;
begin if b = 0 then gcd := a else gcd := gcd(b, a mod b); end;

type
   pnode = ^node;
   node  = array[0..MAXK] of int64;

var
   komb            : array[0..MAXN, 0..MAXK] of int64;
   povrh           : array[0..MAXK] of int64;
   nodes           : array[0..3*MAXN] of node;
   allocator, root : pnode;
   n, k, i, j      : longint;
   h               : array[0..MAXN] of longint;
   gore            : array[0..MAXK] of longint;

procedure calc_povrh(n : longint);
var
   i, j, g, dole : longint;
begin
   povrh[0] := 1;
   i := 1;
   while (i<=k) and (n>=i) do begin
      gore[i-1] := n-i+1;

      dole := i;
      for j:=0 to i-1 do begin
         g := gcd(gore[j], dole);
         dole := dole div g;
         gore[j] := gore[j] div g;
      end;

      povrh[i] := 1;
      for j:=0 to i-1 do povrh[i] := povrh[i] * gore[j] mod MODC;
      
      inc(i);
   end;

   while i<=k do begin
      povrh[i] := 0;
      inc(i);
   end;
end;

function rec(a, b, h0 : longint) : pnode;
var
   curr, up, left, right  : pnode;
   minh, minh_index, i, j : longint;
   leaf                   : boolean;
begin
   curr := allocator; inc(allocator);
   leaf := true;
   minh := 1000000000;
   minh_index := -1;
   for i:=a to b-1 do begin
      if h[i] < minh then begin
         minh := h[i];
         minh_index := i;
      end;
      if h[i] <> h[a] then leaf := false;
   end;

   if leaf then begin
      calc_povrh(minh-h0);
      for i:=0 to k do
         curr^[i] := komb[b-a, i] * povrh[i] mod MODC;
   end else if minh > h0 then begin
      up := rec(a, b, minh);
      calc_povrh(minh-h0);

      for i:=0 to k do begin
         curr^[i] := 0;
         j := 0;
         while (j <= i) and (b-a-j >= 0) do begin
            curr^[i] := curr^[i] + up^[j] * povrh[i-j] mod MODC * komb[b-a-j, i-j] mod MODC;
            inc(j);
         end;
         curr^[i] := curr^[i] mod MODC;
      end;
   end else begin
      left := rec(a, minh_index, h0);
      right := rec(minh_index+1, b, h0);

      for i:=0 to k do begin
         curr^[i] := 0;
         for j:=0 to i do 
            curr^[i] := curr^[i] + left^[j] * right^[i-j] mod MODC;
         curr^[i] := curr^[i] mod MODC;
      end;
   end;
   rec := curr;
end;

begin
   read(n, k);
   for i:=0 to n-1 do read(h[i]);

   for i:=0 to n do begin
      komb[i, 0] := 1;
      for j:=1 to i do
         komb[i, j] := komb[i, j-1] * (i-j+1) mod MODC;
   end;

   allocator := @nodes[0];
   root := rec(0, n, 0);
   writeln(root^[k]);
end.
